compiler/rustc_mir/src/transform/large_enums.rs

```diff
@@ -0,0 +1,218 @@
+use crate::transform::MirPass;
+use crate::util::patch::MirPatch;
+use rustc_data_structures::stable_map::FxHashMap;
+use rustc_middle::mir::*;
+use rustc_middle::ty::{self, Const, List, Ty, TyCtxt};
+use rustc_span::def_id::DefId;
+use rustc_target::abi::{Size, Variants};
+
+/// A pass that seeks to optimize unnecessary moves of large enum types, if there is a large
+/// enough discrepanc between them
+pub struct EnumSizeOpt<const DISCREPANCY: u64>;
+
+impl<'tcx, const D: u64> MirPass<'tcx> for EnumSizeOpt<D> {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        self.optim(tcx, body);
+    }
+}
+
+impl<const D: u64> EnumSizeOpt<D> {
+    fn candidate<'tcx>(
+        tcx: TyCtxt<'tcx>,
+        ty: Ty<'tcx>,
+        body_did: DefId,
+    ) -> Option<(u64, Vec<Size>)> {
+        match ty.kind() {
+            ty::Adt(adt_def, _substs) if adt_def.is_enum() => {
+                let p_e = tcx.param_env(body_did);
+                // FIXME(jknodt) handle error better below
+                let variants = &tcx.layout_of(p_e.and(ty)).unwrap().variants;
+                match variants {
+                    Variants::Single { .. } => None,
+                    Variants::Multiple { variants, .. } if variants.len() <= 1 => None,
+                    Variants::Multiple { variants, .. } => {
+                        let min = variants.iter().map(|v| v.size).min().unwrap();
+                        let max = variants.iter().map(|v| v.size).max().unwrap();
+                        if max.bytes() - min.bytes() < D {
+                            return None;
+                        }
+                        Some((variants.len() as u64, variants.iter().map(|v| v.size).collect()))
+                    }
+                }
+            }
+            _ => None,
+        }
+    }
+    fn optim(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        let mut match_cache = FxHashMap::default();
+        let body_did = body.source.def_id();
+        let mut patch = MirPatch::new(body);
+        let (bbs, local_decls) = body.basic_blocks_and_local_decls_mut();
+        for bb in bbs {
+            bb.expand_statements(|st| {
+                match &st.kind {
+                    StatementKind::Assign(box (
+                        lhs,
+                        Rvalue::Use(Operand::Copy(rhs) | Operand::Move(rhs)),
+                    )) => {
+                        let ty = lhs.ty(local_decls, tcx).ty;
+                        let (num_variants, sizes) = if let Some((nv, s)) = match_cache.get(ty) {
+                            (*nv, s)
+                        } else if let Some((nv, s)) = Self::candidate(tcx, ty, body_did) {
+                            // FIXME(jknodt) use entry API.
+                            match_cache.insert(ty, (nv, s));
```

Comment:
This could cache an interned allocation.

Reply:
would that be a new function on `TyCtxt`? Or is there something already for that?

Reply:
The `Vec` with variant sizes is used to create an allocation later on. The idea was to cache the result of `intern_const_alloc` instead instead of cacheing the `Vec` and recreating the allocation each time.